Implement quality treatment for asynchronous actions in multicore (3/3?)

gadmm opened the pull request:
Here is for your consideration a simplification of the remaining commits at #11057 (after #11095 and #11190).

This PR includes various fixes for soundness and performance of async actions.

In particular there is a bug whereby `caml_enter_blocking_section` can loop waiting for another domain to process its signals (since `caml_check_pending_signals` is shared between all domains). The PR #11057 proposed two different ways to fix this:
- One was to introduce a distinction between "external" and "internal" interrupts (https://github.com/ocaml/ocaml/pull/11057/commits/4fbd1225ec at #11057). (_External Interrupts_)
- One is to delay the setting of `young_limit` in case action are pending until one returns to OCaml code (this PR, https://github.com/ocaml/ocaml/pull/11307/commits/93dd8f28db). (_Optimal Polling_)

Optimal Polling fixes more things than External Interrupts, and while the resulting implementation is simpler to understand, it is more invasive (e.g. changes to `.S` files). So I originally intended it for a part 4/4 of this PR series not needed for 5.0.

While rebasing this PR, I realised that the simplifications Optimal Polling brought were greater than thought (also, there might be a bug in the current implementation of External Interrupts, so I would have to look at it again). So I decided in a first time to skip the External Interrupts design, so that you do not have to review two separate designs (where the second one undoes the changes of the first one), and go directly for the "nice" design. (For explanations of the "nice" design, see the lengthy documentation comments and commit logs.)

If you prefer starting with External Interrupts first, then we can delay the review of Optimal Polling until 5.1. If you like the Optimal Polling approach, then I will write a test to ensure that any future architecture implements the necessary code when returning from C to OCaml.

Commit logs have been updated to reflect previous discussions (cc @xavierleroy).

The commits are best reviewed separately. In case of the "Optimal Polling" commit, looking at the resulting implementation (and checking that it matches the description in the comments) might make more sense than looking only at the diff.

(cc @sadiqj, @gasche, @kayceesrk.)

gadmm committed:
Optimal polling in C code based on masking

Do not slow down allocations from C when an asynchronous callback
cannot be handled immediately.

Before this commit, an async action pending in C code, and thus not
immediately processed inside allocations, is remembered by leaving
[young_limit] to the max value after the allocation. Thus all
subsequent allocations go through the slow path until the action is
processed. There is also an [action_pending] flag which is a bit
redundant in remembering that an async action is pending.

This commit introduces a simpler design whereby:

- [action_pending] alone is used to remember that an async action is
  pending,
- [young_limit] is only used to interrupt running code in order to
  notify of new pending actions.

When switching from C to OCaml, the value of [action_pending] is used
to set [young_limit] for interruption as necessary. This is done when
returning from C to OCaml (see RET_FROM_C_CALL) or when calling into
an OCaml callback (see caml_update_young_limit_after_c_call).

Other relevant changes:

- We can always reset [young_limit] to its regular value inside C code
  and thus avoid the repeated slow path.
- We make [action_pending] a _Bool (better code generation for
  RET_FROM_C_CALL across platforms)
- [caml_set_action_pending] no longer immediately interrupts the
  program; the [action_pending] flag is no longer used in this way.

This new design can also be seen as a step towards implementing
masking (i.e. delaying async action inside OCaml code blocks just like
inside C code).

gadmm committed:
Do not access the TLS from signal handlers

When a signal arrives, we interrupt all domains. This is a strategy
that makes much fewer assumptions about signal-safety, threads, and
domains. The point of view of this commit is that realistic uses of
signals (SIGINT, SIGALRM...) make them arrive infrequently-enough that
this is affordable.

* In mixed C/OCaml applications, there is no guarantee that the POSIX
  signal handler runs in an OCaml thread so [domain_state] might not be
  available.

* While C11 mandates that atomic thread-local variables are
  async-signal-safe for reading, gcc does not conform and can allocate
  in corner cases involving dynamic linking.

* It is unclear whether the OSX implementation of [domain_state] is
  safe to read from signal handlers either, but this might be a
  theoretical concern only.

* Do not make the hypothesis that the thread executing a POSIX signal
  handler is the most ready to execute the corresponding OCaml signal
  handler (e.g. Ctrl-C sent to the toplevel received by domain 0 when
  stuck inside [Domain.join]).

* These changes allow simplifications to the spawning of threads and
  domains since we no longer need to mask signals during thread
  creation. (See subsequent commit.)

The biggest risk of this change is the fact that C code allocating on
the GC heap repeatedly polls without clearing pending actions until a
safe point is found. By interrupting all domains when a signal
arrives, the cost of this phenomenon is amplified. A subsequent commit
fixes this behaviour by making sure that a signal notified to a domain
causes this domain to poll only once inside C code.

gadmm committed:
Do not raise from `caml_thread_yield`

gadmm committed:
Simplify spawning of threads and domains

As follows from the preceding commit.

gadmm commented on runtime/signals.c:
```diff
@@ -103,11 +103,34 @@ CAMLexport value caml_process_pending_signals_exn(void)
+   processed in caml_alloc_small_dispatch.
+
+   We interrupt all domains when a signal arrives. The point of view
+   of this commit is that signals (SIGINT, SIGALRM...) arrive
```
```suggestion
   of this approach is that signals (SIGINT, SIGALRM...) arrive
```

gadmm commented on runtime/amd64.S:
```diff
@@ -535,6 +535,15 @@ ENDFUNCTION(G(caml_allocN))
+        ret;                                    \
+1:                                              \
+        movq    $-1, Caml_state(young_limit);   \
+        ret
```
For a start, adding this to the .S is only one way of implementing it. I looked around for other ways and it seemed the simplest to me, but maybe you know better ways. Notice that we only need it at the end of non-noalloc C calls, so for easier portability it might cheap enough to call `caml_update_young_limit_after_c_call` instead (if there is a good place to insert such a call).

abbysmal replied:
I cannot think of a nicer way at the moment, it seems reasonable to me

kayceesrk replied:
1. What happens if you don't update the young limit here? Will the signal be delayed arbitrarily or forgotten? Is it a matter of timely handling of the pending action or is it a matter of correctness?  
2. Needs to be implemented for the newer backends. RISC-V, s390x, WIP (Power) @xavierleroy.

xavierleroy replied:
I still need to understand what this code is trying to do.  Explain me like I'm 6.

gadmm replied:
@xavierleroy An explanation of the new `action_pending` mechanism is given at [signals.c#R281-R305](https://github.com/ocaml/ocaml/pull/11307/files#diff-93ee9a0b40685be35da1909f57cba82cfe51d5a8de495947e5068a8b8d8172cbR281-R305). The explanation is intended to be more complete and self-contained than the text it replaces, thanks to this approach being simpler. Let me know if this text is helpful; if unclear I can try to clarify it, and if still unclear I have a metaphor for 6-year olds (involving a postman and a flag on the mailbox) that I can write down when time allows.

gadmm replied:
@kayceesrk The handling of actions can be delayed until the next minor collection, if we fail to adjust the young_limit in accordance with the action_pending flag.

gadmm replied:
@kayceesrk I thought about it further—if the next minor collection also happens at an allocation from C, then the callbacks are delayed again until the next minor collection, and so on. So although unlikely, they can in theory be delayed an arbitrarily long time (e.g. code that spends much time on the C side). Also, the minor collection could also fall while another thread is scheduled, which in theory can again delay callbacks an arbitrarily long time (e.g. if the thread that must run an async callback allocates very little compared to the other threads).

kayceesrk replied:
Thanks for the explanation. 

The code improves on the minor possibility that asynchronous actions may be arbitrarily delayed. This is done by checking that there are pending asynchronous actions when returning from an C call. If so, the young limit is set to a large value so that the next allocation in OCaml fails. 

Benchmarking previously done has shown that this does not add a significant overhead.

gadmm replied:
This, and also the delaying that would happen for sure would be detrimental for memprof callbacks which should run as soon as possible.

kayceesrk replied:
@xavierleroy are you satisfied with the responses in this thread?

xavierleroy replied:
I would have loved to read the story with the postman and the mailbox :-)   But, yes, the comments in the code clearly describe the design and the implementation.  Thanks!

The only question that popped in my mind (but I think I have an answer) is this: when we're about to return from `caml_c_call` and notice that `action_pending` is set, why don't we jump directly to C code that can handle the pending action, instead of setting `young_limit` and relying on the next poll point in the OCaml code to notice and handle the pending action?

One possible answer is that the frame descriptor that ocamlopt attached to the C call site is not quite right for this purpose: the return value of the C call would have to be kept alive specially, and the backtrace may be wrong if the delayed async action raises an exception.

Another answer is that it would need more code in the runtime/*.S files, something we should avoid for our sanity.

So, I have no further questions on this PR, just one pedantic assembly code remark that can be ignored (next comment).

gadmm replied:
This amounts to adding a polling location when returning from C code. If that's what we want, we can do it, but there is always a danger for code that is careful to control polling locations (especially in a place where that is likely to return a resource). For instance this would break the resource-safety property of `Mutex.protect` https://github.com/ocaml/ocaml/blob/trunk/stdlib/mutex.ml#L27 as it stands, and it would be more work to restore this property.

kayceesrk replied:
Thanks for the response. The assembly code remark has been fixed in https://github.com/ocaml/ocaml/pull/11307/commits/53dedf18eb79e8f2dbc3edf66e017b2715f7def7. I shall merge this PR.

abbysmal commented:
Hey!
Thank for this PR. Is it ready for review at the moment?
I am willing to take a look at it, at least for the parts I can understand well enough.

gadmm commented:
Yes, it is ready for review. I consider it in polished state after several rounds of self-review. I have not followed the 5.0 alpha business but given the fixes it brings (and the fact that it is a pre-requisite for #11272), it is intended for 5.0.

Octachron commented:
Concerning the alpha business, all fixes are still in scope for 5.0 during the alpha and beta release phase. Please don't hesitate to ping me on PRs that should be integrated in 5.0 once reviewed.

gadmm commented:
I also saw you mention that for this "zero" release, C runtime API can still change, so I can ping you for things like #11272 and things related to #10992, do I understand correctly?

Octachron commented:
Indeed, those kind of known possible C runtime API changes are one of the reason we have this 0th alpha release with less stability guarantees than an ordinary alpha.

gadmm commented:
@sadiqj @gasche @kayceesrk This fixes bugs for 5.0 and it is a pre-requisite for #11272. Is anyone planning to have a look?

kayceesrk commented:
> Is anyone planning to have a look?

Not this week as I am busy with other things.

abbysmal commented on runtime/domain.c:
```diff
@@ -1462,10 +1473,23 @@ int caml_try_run_on_all_domains(
+    atomic_uintnat * interrupt_word =
+      atomic_load_explicit(&d->interruptor.interrupt_word,
+                           memory_order_acquire);
+    if (interrupt_word == NULL) return;
```
Instead of looping over `all_domains`, wouldn't it be simpler to just check for every domains that are part of the STW participating set?
This may be a naive suggestion, there may be some odd cases where it is important to check for `interrupt_word` nonetheless, but to me checking the participating set solves both having to check for `interrupt_word == NULL` and doing a for loop over `all_domains`.

gadmm replied:
There is a problem of synchronisation. The difference is that the participating set is an array of pointers, whereas all_domains is an array of values (with some monotonicity property). So you can access interrupt_word directly with all_domains, without synchronisation (other than with other people who access interrupt_word). With the participating set you cannot lock the mutex in a signal handler, so you would need to rewrite the manipulations to the participating set for reading without a mutex.

I also do an early exit to avoid looping over all_domains until the end: if interrupt_word is null then we reached the end.

abbysmal commented on runtime/domain.c:
```diff
@@ -1462,20 +1467,37 @@ void caml_interrupt_all_for_signal(void)
-      || atomic_load_relaxed(&dom_st->requested_external_interrupt)
-      || dom_st->action_pending) {
-    atomic_store_rel(&dom_st->young_limit, (uintnat)-1);
-    CAMLassert(caml_check_gc_interrupt(dom_st));
```
If I understand correctly, `interrupt_domain_local` in the change will still make the assertion you removed valid?
Was it removed because of relative usefulness?

gadmm replied:
The story is that I added this assert after being bitten by an off-by-one bug. This concerned more the "External interrupts" approach where young_limit was not always increased to UINTNAT_MAX. The "Optimal polling" approach definitely settles for UINTNAT_MAX everywhere so I am not worried about the bug reappearing by mistake.

abbysmal commented:
Regarding review:
I was currently attempting to wrap up mine, but it seems Github got lost on your last force push and lost half of my comments.
Sorry for the noise, I will wrap it up properly shortly.

gadmm commented:
Very sorry about this. I was not aware of this effect of force pushes. Are they not accessible from the old branch https://github.com/ocaml/ocaml/commit/d549824d6a4e1026643feaead4b0dceeb99c891b ?

abbysmal commented:
I think it had to do with how the Github UI records comments when you do a review commit by commit, force-push probably messed with this, I know now not to trust this too much. :-)

abbysmal commented on runtime/amd64.S:
```diff
@@ -535,6 +535,15 @@ ENDFUNCTION(G(caml_allocN))
+        ret;                                    \
+1:                                              \
+        movq    $-1, Caml_state(young_limit);   \
+        ret
```
Double ret should maybe be avoided here.

gadmm replied:
I think this is the best implementation: forward jumps are predicted not-taken in the absence of branch predictor history, so it is guaranteed that it costs (almost) zero in the fast path. This is the code generated by gcc (or clang, I don't remember; I used godbolt) after I using the appropriate `__builtin_expect`. (I will add comments to this effect to facilitate the implementation in future targets.)

abbysmal commented on runtime/signals.c:
```diff
@@ -248,22 +274,30 @@ void caml_request_minor_gc (void)
+   Asynchronous actions are notified to the domain by playing with the
+   allocation limit. Non-delayable actions are performed immediately,
+   then [Caml_state->action_pending] is set in order to record that an
+   action of the second kind might be pending. Then those actions are
```
I think this part is a bit confusing, it took me a few attempt to tie together the paragraph.
It may have to do with using the `second kind` term to refer to delayable actions?
I understand it fine looking at the code and re-reading, I'm unsure how to rewrite it better.

gadmm replied:
I have reworded a bit

abbysmal commented on runtime/signals.c:
```diff
@@ -141,16 +163,20 @@ CAMLexport void (*caml_leave_blocking_section_hook)(void) =
-    if (!caml_check_pending_signals()) break;
+    /* Check again if a signal arrived in the meanwhile. If none,
+       done; otherwise, try again. */
+    if (atomic_load_relaxed(&domain->young_limit) != UINTNAT_MAX) break;
```
I am not very fond of spreading more the (relatively) internal knowledge that `young_limit == UINTNATMAX` means interrupt.
In some places it can't be avoided (runtime assembly changes), in this case maybe a macro would fit better.
Especially since the comment suggests that `young_limit == UINTNAT_MAX` means a signal arrived: it could be in fact a minor/major cycle?
There's also a slight distinction between `Caml_check_gc_interrupt` and checking for `UINTNAT_MAX`: the former has a broader meaning (could be that we reached the end of the minor segment, or we hit an interrupt).
I do think this code segment is fine, but the presence of two form to check for interrupts looks a bit strange to me, maybe I am missing a subtlety.

gadmm replied:
It might not be a signal (it might be a cross-domain interrupt), but this is fine.

I'll need to have a look to see why I decided not to use Caml_check_gc_interrupt

gadmm replied:
The reason is simple: one can call `Caml_check_gc_interrupt` only when holding the domain lock, since it wants to read `young_ptr`. `UINTNAT_MAX` just appeared to be a convenient value to compare to, there is no tricky mechanism hidden there.

abbysmal reviewed (commented):
Thank you for your PR!

I had a go at reviewing this PR, and left a few questions and nitpicks.
I am not quite familiar with the newer advances in the design of signals handling in the runtime, another look (by someone with more recent knowledge :-)) will be required.

The approach does make sense to me, I think it does as advertised and has the nice property of eradicating masking in both domain and systhread bootup.

I will need to re-read a few sections, I have more questions coming for `caml_interrupt_all_for_signal`, but I need to document myself a bit more first.

gadmm commented:
Is there anything blocking this delicate bugfix PR for 5.0?

gadmm commented:
@gasche Can you please put a 5.0 milestone on this or something?

xavierleroy commented:
I could only have a very quick look, but your "optimal" strategy is already making me nervous performance-wise.  `caml_c_call` is a hot path in ocamlopt-generated code, so adding a load and a test there can have significant costs.  A round of Sandmark bechmarking is needed before we can go any further in this direction.

Remember that 99% of OCaml programs do not handle signals.  I don't want to penalize them to avoid corner cases in the remaining 1%.  We've already paid enough with the polling points.

gadmm commented:
I'll look into having benchmarks for this PR.

The reasoning performance-wise is that the additional check involves a correctly-predicted branch, which can be free, and a load from a cache line that is already accessed in the same code. In addition the check is not needed (and not implemented) after "noalloc" C calls.

(This does not affect this PR, but I do not think that the "1%" figure is a good estimation to have in mind regarding asynchronous callbacks: 1) we have already seen that interrupts are fairly commonly used in some application domains dear to OCaml, and 2) if we keep pathological corner-cases then users of memprof and finalizers are also affected even when those callbacks do not raise.)

gadmm commented:
(Temporarily reverting fd3db5e for benchmarking purposes.)

gadmm commented:
Here are results of sequential benchmarks:

Turing host:
![turing](https://user-images.githubusercontent.com/19315355/187096446-208e0301-850d-490f-9355-7771dba9730a.png)

Navajo host:
![navajo](https://user-images.githubusercontent.com/19315355/187096453-8e8677f4-27e0-43e6-a66e-bae770e1746b.png)

I do not see anything too striking in these results, but I am not very familiar with this benchmark suite. The turing one looks favourable but I do not have an interpretation to offer for the good results.

I am running additional benchmarks to see if the sensitivity of lwt-based benchmarks is related to fd3db5e (new unconditional polling after every blocking section). These lwt benchmarks look like micro-benchmarks, so such a slowdown (if actually caused by this change) might not be representative of real programs.

gadmm commented:
I am considering withholding the commit "Check actions unconditionally when leaving a blocking section" for now. There is a chance that it affects the lwt micro-benchmarks and although I do not think there is an issue with it, it is not needed as a bugfix.

Here are the new results without this commit:

Turing:
![turing](https://user-images.githubusercontent.com/19315355/187532476-6fb01c00-0b4e-4cc2-8e2e-8ff61a700c52.png)
Navajo:
![navajo](https://user-images.githubusercontent.com/19315355/187532636-64065e88-c154-4f04-9d44-81bf39e139d4.png)
(note that there are similar variations on navajo for levinson_durbin, sequence_cps and other bechmarks when comparing trunk with itself as well; this is not related to this PR)

Withing the limits of sandmark benchmarks, the results seem favourable.

gadmm commented:
Rebased & history cleaned-up

Octachron commented:
Since the change is quite invasive, I am very tempted to move this PR to the 5.1 milestone to give more time to the reviewers and alternative designs.

gadmm commented:
I do not really understand what is blocking this PR. The original PR was split in several PRs after people asked for it, the series of PRs was not designed to be merged in separate releases (except where otherwise indicated). Also I am not aware of alternative designs being proposed at this point, and this is the only one that relies on a well-defined mechanism as opposed to playing poke-a-mole with ill-behaving corner-cases.

gadmm commented:
Re: benchmarks.

@kayceesrk suggested that I look at the newly added `perf` output in sandmark nightly. I looked at it a while ago but there was not much to conclude from it. Some benchmarks show a noticeable instruction count increase, e.g. 1% for quicksort. This turned out to be due to calling the generic C comparison function `caml_lessthan` on integers, and thus making so many short-lived C calls that the 2 instructions added when returning from C calls amounted for this 1% extra instruction count. The conclusion is that this benchmark is far from representative of actual code. Despite this, there does not seem to be a clear measurable slowdown: the current sandmark runs show a slowdown, the earlier runs as well as my own tests show a speedup (for one particular layout of code, etc.).

I think it does not make sense to request sandmark benchmarks for things like adding a well-predicted branch with a short condition in a couple of locations. At the OCaml workshop I showed that one can add a well-predicted branch inside the hottest loop of the runtime and see no significant impact on performance, yet to attest this required far more precise and time-consuming measurements than what sandmark gives us.

gasche commented:
Shortly:

1. What is "blocking" the PR. Our reviewer time is limited by human factors and this PR requires more review before being mergeable. On the other hand, Florian's job as our release manager is more or less to get 5.0 out as fast as reasonably possible. In this context, we don't want this PR "blocking" things. 

   Reviews on this PR are not "blocked", they can happen whenever independently of the release schedule, it just looks like everyone is working on something else right now that they assume has higher priority. Note that in currently the runtime maintainers do not consider "asynchronous actions" a priority. (I believe that it is important at least for statmemprof, but statmemprof support is not expected in 5.0. You fixed bugs in the Threads implementation, and this very nice, but we could possibly live with a couple remaining bugs for 5.0, there will be more releases. Personally I think the most important aspect is that we collectively shared a better understanding of the Threads codebase -- and the fact that several pain points remain there.)

   > the series of PRs was not designed to be merged in separate releases

   What are the risks with having a release in-between?

2. I don't know why you consider that short-lived C calls are "not representative of actual code", to my knowledge they are the only portable way to access C-supported intrisics from OCaml (say popcount, etc.), so they could easily occur in performance-critical loops. It is only representative of *some* actual code, not of most existing code, but having some benchmarks detecting this in Sandmark sounds reasonable. (I have no opinion over whether this particular benchmark should do this. Did you open an issue against Sandmark to discuss this?)

3. Even when Sandmark is not accurate enough to analyze the performance impact of a change, people get reassured by the pretty pictures showing white noise. Don't laugh! It's worth running Sandmark if only to show that there is no noticeable *large* performance impact.

gadmm commented:
Thanks for the reply @gasche. I am aware that OCaml 5.0 is expected to be buggy, I was more worried about the mention of alternative designs, as I am not aware of any.

1. The first commit deals with hard-to-reproduce crashes in corner cases. A large number of packages define signal handlers for various signals (sometimes to raise exceptions, but not always), but it is hard to be sure that any will be affected. I am most worried about signals coming when `domain_state` is not available (e.g. inside a non-ocaml thread spawned by foreign code) but it is hard to tell how serious it is, in part because I am not very familiar with how such situations arise. (This would have been UB even before multicore, but accidentally working in practice, because the previous C signal handler was not thread-safe either. Since this probably used to work in practice, we cannot use this to rule out that the situation exists.)

   The third commit fixes a situation where it is still possible to have an asynchronous exception arising when allocating from C code. This is contrary to the current bold claims in the documentation, but it was unnecessary to propose a separate fix to the documentation with a better fix in the pipeline.

   The second commit is cosmetic and the fourth one addresses a theoretical performance concern about the strategy of the first commit (it is also a step towards a simpler implementation of masking, but this is not important here). I would not have proposed the first commit without the fourth one, but the fourth can be omitted for 5.0.

   I can propose the first commit and the third commit for 5.0, which should be seen as less invasive than the whole PR. (In hindsight, I could have proposed this a while ago, given better communication about the release schedule of OCaml 5.0 and its expected state when released.)

2. Note that `noalloc` C calls are not affected by the change (so this excludes `popcnt`, etc.).* Anyway the lesson is that with an unrealistic proportion of instructions taken by OCaml-to-C switching, we could still not distinguish a significant impact.

   (*unrelated: IIUC, `caml_compare` is non-noalloc because it needs to raise exceptions sometimes, and this gave me the idea that, if performance is important in this particular case (with or without this PR), it would be possible to let `noalloc` calls raise exceptions in a new calling convention where any exception raised is returned encoded; in other words making a `noalloc` call to some `caml_compare_exn` function.)

4. It is fine and natural to ask a benchmark for the PR as a whole to see if there is not an impact as you describe (we have such results, and they show no such large impact, so far so good). This is missing my point. Sandmark cannot let you see if there is an impact of adding a well-predicted branch in a hot location in the code. It is like trying to use a magnifying glass to study bacteria. In this case, sandmark might randomly be slightly skewed one way or another in a repeatable way (because it goes out of its way to repeat the same random spot in the possible outcomes). As for instruction count, a large increase then turns out to be due to the micro nature of the benchmarks (and is not a good measure when the argument for good performance is increased IPC). This has none of the reassuring effect you seek and wastes a lot of time in trying to interpret non-existent effects; it also incites to be overly cautious (as I think I was for #11506 due to sandmark).

gadmm committed:
Changes

gadmm commented:
This is now rebased.

The commits descriptions have been updated and clarified, hoping this helps. Nothing has changed in the code.

gadmm committed:
[minor] Better documentation for catch_break

As implemented by the previous bugfix commit.

gadmm commented:
The first commit (interrupt all domains inside the C signal handler) fixes the following testcase. This is an example where the domain receiving the signal is not the most able to handle it.
```ocaml
let break_trap s =
  (try while true do () done
   with Sys.Break -> print_endline "[Sys.Break caught]" ) ;
  print_endline s

let () =
  Sys.catch_break true ;
  let d = Domain.spawn (fun () -> break_trap "Still alive") in
  break_trap "There is research to be done." ;
  Domain.join d ;
  break_trap "We do what we must because we can." ;
  print_endline "I'm making a note here: Huge success."
```
In trunk, Ctrl-C will interrupt the first call to `break_trap`. It will hang on the second call to Ctrl-C: SIGINT is received by domain 0, which is stuck on Domain.join, whilst domain 1 is stuck in a loop and is never interrupted no matter how many times you press Ctrl-C. (You need to `kill -9` the program.)

(You can remove `Sys.catch_break true` and try it in the toplevel, it also "works" in bytecode.)

With the present PR, you get an interaction like the following:
```
^C[Sys.Break caught]
Still alive
^C[Sys.Break caught]
There is research to be done.
^C[Sys.Break caught]
We do what we must because we can.
I'm making a note here: Huge success.
```

The theory says that someone using threads and signals together will use masks to direct the signal to the right thread/domain. However, beyond the toplevel, one can imagine that for simple script-like usage the programmer will simply call `Sys.catch_break true` and be happy to interrupt the program by pressing Ctrl-C enough times. Incidentally, it is not a fabricated example: I found it by chance while working on another issue.

I am adding a test to the suite inspired by this example.

xavierleroy commented:
After all this time, it's still not clear to me what the performance impact of these changes are on programs that do not handle signals, which I still think as the overwhelming majority.  Could you recapitulate this for me?  I'm not asking for more measurements yet, just a recap of your intuitions on what could be a performance impact / very likely has no performance impact.  Thanks.

gadmm commented:
Assuming a program either does not set up OCaml signal handlers, or does not receive signals (but maybe uses other async callbacks: finalisers, memprof callbacks):

1. "Do not access the TLS from signal handlers":
   - No impact (e.g. replace a plain read by a relaxed atomic read)
2. "Simplify spawning of threads and domains":
   - No impact (remove code that was only useful for users of signals)
3. "Do not raise from caml_thread_yield":
   - ~Correctness issue: fixes a bug whereby other kinds of callbacks could raise in those locations, but should not.~ Edit: no impact.
4. "Optimal polling in C code based on masking"
   - Potential negative impact: add a load+test+branch in two locations (the end of `caml_c_call` and `caml_c_call_stack_args`): ruled negligible (cache-local data & statically well-predicted branch)
   - Potential negative impact: add calls to `caml_update_young_limit_after_c_call` inside callback.c: idem; in addition, the rest of function bodies looks comparatively expensive, so it did not seem worth it to make the function call inlinable
   - Potential negative impact: add calls to `caml_update_young_limit_after_c_call` in interp.c: idem; in addition the cost looks even more negligible inside the interpreter 
   - Positive impact: avoids taking the slow allocation path in C code if other kinds of callbacks than signal handlers are pending (or are believed to be pending: OCaml forces the examination of all callbacks twice per minor cycle, which will also cause these repeated slow paths)
   - Potential positive impact: guard the function call to `caml_process_pending_signals_exn` inside of `caml_enter_blocking_section` (negligible)

avsm commented:
triage meeting: @kayceesrk to review PR and look for a minimal change for 5.1 to fix #12253, with the remainder of fixes targetted at trunk.

gadmm commented:
Thanks @kayceesrk for having a look. I have seen your question about a minimal fix for 5.1; I'll answer shortly.

kayceesrk commented:
> Thanks @kayceesrk for having a look. I have seen your question about a minimal fix for 5.1; I'll answer shortly.

Thanks a minimal fix would be incredibly useful.

gadmm commented:
The fix for the two examples (#12253 and https://github.com/ocaml/ocaml/pull/11307#issuecomment-1407524610) is given by the 1st and 4th commits. The 4th commit fixes the example at #12253 but was designed with the 1st one in mind (see below). The 1st fixes the other example but the 4th was proposed to alleviate potential performance reasons, and because the original intermediate approach without the 4th commit turned out to be flawed (as explained at the top).

The fix for #12253 happens here:
```c
     /* Check again for pending signals.
       If none, done; otherwise, try again */
    // FIXME: does this become very slow if a signal is recorded but
    // is masked for everybody in capacity of running signals at this
    // point?
    if (!caml_check_pending_signals()) break;
```
being replaced by:
```c
    if (atomic_load_relaxed(&domain->young_limit) != UINTNAT_MAX) break;
```
This is a weaker test. But with the 1st commit, which interrupts all domains when a signal arrives, it operates essentially like the old `!signals_are_pending` test from OCaml 4.

To the best of my understanding, the test above participates in a _best effort_ to reduce the risk of race before entering an interruptible syscall. Hence if looking to fix only #12253 for 5.1, doing just this one-line change sounds sensible. You also need the modification right above it in the 4th commit, to reset the young limit at the start of the loop, otherwise you have an infinite loop[^1]. However, it is not clear to me that this would work as intended on its own. Such a potential fix would have to be elaborated and reviewed in its own PR. What makes it hard is that the previous implementation (prior to this PR), as I now understand, has no clear specification and easily creates unexpected issues that requires to have the full implementation in mind—and I am afraid at this point I have forgotten its subtle pitfalls.

@kayceesrk I understand where you are coming from with your request but it was very time-consuming—investigating counterfactuals such as "would this commit make sense without such and such commit" or "would such and such change make sense without the rest of its commit" are not questions that otherwise arise during review and they are difficult to answer, especially a year later. The usual practice of the OCaml community with fixed releases to not rush anything prior to release and merge PRs as they are ready is a good one to follow here.

You have criticised elsewhere this PR for its readability, but I do not see how your suggestions apply here. From the start I have dedicated significant effort to make this PR into clean separate commits with detailed change logs. The parent PR has already been split into 3 parts following reviewer's requests, despite being originally meant to be released all at once. Since then I have not received further feedback on the format of this PR. I acknowledge that the 4th commit is harder to review from the diff only, in fact I underlined this at the beginning of the PR.

It is good to take a step back and consider that we are reviewing some code that is meant to complete an original implementation in multicore of async callbacks which was actually out of sync with OCaml 4 when merged and which itself did not receive so many comment and requests (apart from my own).

[^1]: to anticipate on an obvious remark, note also that it is tempting to simply call `caml_process_pending_actions` at the start of the loop—the decision not to do so was discussed in another PR a long time ago.

gadmm commented:
I went and drafted what I thought about here:
```diff
diff --git a/runtime/signals.c b/runtime/signals.c
index f33e1cc47c..ec8cbf3c30 100644
--- a/runtime/signals.c
+++ b/runtime/signals.c
+++ b/runtime/signals.c
@@ -144,16 +144,21 @@ CAMLexport void (*caml_leave_blocking_section_hook)(void) =
 
 CAMLexport void caml_enter_blocking_section(void)
 {
+  caml_domain_state * domain = Caml_state;
   while (1){
     /* Process all pending signals now */
-    caml_raise_if_exception(caml_process_pending_signals_exn());
+    if (Caml_check_gc_interrupt(domain)) {
+      Caml_state->action_pending = 0;
+      /* Reset young_limit. */
+      caml_handle_gc_interrupt();
+      caml_raise_if_exception(caml_process_pending_signals_exn());
+    }
     caml_enter_blocking_section_hook ();
-    /* Check again for pending signals.
-       If none, done; otherwise, try again */
-    // FIXME: does this become very slow if a signal is recorded but
-    // is masked for everybody in capacity of running signals at this
-    // point?
-    if (!caml_check_pending_signals()) break;
+    /* Check again if a signal arrived in the meanwhile. If none,
+       done; otherwise, try again. Since we do not hold the domain
+       lock, we cannot read [young_ptr] and we cannot call
+       [Caml_check_gc_interrupt]. */
+    if (atomic_load_relaxed(&domain->young_limit) != UINTNAT_MAX) break;
     caml_leave_blocking_section_hook ();
   }
````
By writing it, I see that you forget other pending callbacks, and you must reset `action_pending` (otherwise it might loop). Note that in this case this is a matter of running finalisers now or forgetting them. Calling `caml_process_pending_actions` instead might be a solution; the decision not to call `caml_process_pending_actions` was made when the callbacks also included those of memprof so the decision might be different here.

(Edit: the above patch causes many tests to time out, I do not immediately see why.)

kayceesrk commented:
Thanks for the response. 

I'll respond to the comment about the organisation of the PRs. I'll look at the technical bits a little later. 

I can understand the frustrations. We appreciate the effort that you put into tricky parts of the runtime system and the effort to split the original work into 3 PRs. I am mainly trying to address this problem that you point out:

>  Since then I have not received further feedback on the format of this PR.

> which itself did not receive so many comment and requests (apart from my own).

I am trying to communicate to you the difficulty that reviewers have found with reviewing the PR. We do not have many reviewers familiar with the parts of the runtime system your PR touches. And they are quite tricky parts. My goal is to ensure that the fixes go in. To this end, I am trying to help with suggestions that may make the PR easier to review so that we not only get reviews from experts in these parts of the code but also others who may not be familiar with these parts but want to learn and contribute by reviewing code.

gasche commented:
I have also discussed with @gadmm the difficult of reviewing some of his PRs. My current thinking is that my main criticism is the length of the PR / the number of commits. Reviewing one of those PRs (even the current one which is itself split from a larger PR) takes effort and time, too much for casual reviewers (people who review to help but are not focused on a specific PR/feature/fix). So my current recommendation for you @gadmm would just be to sent PRs in smaller bursts of commits, as small as reasonably possible, so that we can review each part quickly and make progress. My target would be a 15mn review session by PR. The more technical/subtle the change, the more time we need per-commit, so we should aim for very short PRs.

This being said: this is just an intuition, and I'm not sure that my description of the issue (which is a social/human one) is accurate. Some of our former recommendations have required substantial effort on @gadmm's side while providing only moderate benefits in reviewing/integration speed. (We did move to zero to non-zero integration speed, thanks to reviewing effort from @sadiqj mostly and myself.)

gadmm commented on testsuite/tests/parallel/catch_break.ml:
```diff
@@ -0,0 +1,63 @@
+include systhreads
+** not-windows
+*** bytecode
+*** native
```
I see that the format has changed for some reason. Can someone please point me to the new way of writing this?

gasche replied:
Yes, see https://github.com/ocaml/ocaml/blob/24ec1395087c1ed41a233b2e36203725b490a117/testsuite/tests/parallel/domain_serial_spawn_burn.ml for an example. I would try to port your test as follows:

```ocaml
(* TEST
hassysthreads;
include systhreads;
not-windows;
{
  bytecode;
}{
  native
}
*)
```

gadmm replied:
Thanks! (For people learning the new format with me: @gasche's solution only misses the semi-column after `native`)

kayceesrk commented:
First commit that interrupts all the domains on a signal (https://github.com/ocaml/ocaml/pull/11307/commits/f43d0b2a805d0f68a5216a75dedafa97005f2a8c), looks correct. 

Some notes for self: Instead of interrupting every domain, can we not interrupt just the initial domain? We know that the initial domain will always be around. If the initial domain terminates, the program terminates. The advantage is that every domain is not penalised for the arrival of a signal. The downside is that the thread that services the initial domain may be blocked in the OS. It would make ctrl-C to be delayed arbitrarily delayed. This isn't acceptable.

kayceesrk commented:
The second commit is also OK as we no longer need to worry about the domain-local state not being available in a new thread that is being spawned.

kayceesrk commented:
I have questions about the third commit. Why shouldn't `caml_thread_yield` raise an exception? Can you clarify?

kayceesrk commented:
I've left a question on the fourth commit -- https://github.com/ocaml/ocaml/pull/11307/files#r1231753219.

kayceesrk commented:
The rest looks OK.

kayceesrk reviewed (changes requested):
I have a couple of questions. See comments above.

If `RET_FROM_C_CALL` is going to be retained, it needs to be implemented for RISC-V, s390x and the WIP Power backend.

gadmm commented:
I would like to decouple this PR from the 5.1 milestone to take the necessary time to address the new comments.
I thought the [patch above](https://github.com/ocaml/ocaml/pull/11307#issuecomment-1579722946) failed the tests but they no longer fail. This might have been a transient problem on my machine.

There was another issue though, which is that resetting `action_pending` to 0 breaks an invariant of `caml_process_pending_actions`. This is necessary because `action_pending` is still coupled with `young_limit` at this point (the goal of the 4th commit is to decouple them). As a consequence, other callbacks could be forgotten until the next GC.

To make sure that no action is forgotten at the beginning of the loop, one can perform all actions right away instead of only running the signal handlers. The following diff is against trunk and fixes #12253:
```diff
diff --git a/runtime/signals.c b/runtime/signals.c
index f33e1cc47c..23b3cd41e4 100644
--- a/runtime/signals.c
+++ b/runtime/signals.c
@@ -144,16 +144,16 @@ CAMLexport void (*caml_leave_blocking_section_hook)(void) =
 
 CAMLexport void caml_enter_blocking_section(void)
 {
+  caml_domain_state * domain = Caml_state;
   while (1){
     /* Process all pending signals now */
-    caml_raise_if_exception(caml_process_pending_signals_exn());
+    caml_process_pending_actions();
     caml_enter_blocking_section_hook ();
-    /* Check again for pending signals.
-       If none, done; otherwise, try again */
-    // FIXME: does this become very slow if a signal is recorded but
-    // is masked for everybody in capacity of running signals at this
-    // point?
-    if (!caml_check_pending_signals()) break;
+    /* Check again if a signal arrived in the meanwhile. If none,
+       done; otherwise, try again. Since we do not hold the domain
+       lock, we cannot read [young_ptr] and we cannot call
+       [Caml_check_gc_interrupt]. */
+    if (atomic_load_relaxed(&domain->young_limit) != UINTNAT_MAX) break;
     caml_leave_blocking_section_hook ();
   }
 }
```

Now, as I wrote, in previous PRs we were careful not to do this. However, I cannot find the relevant discussion anymore. Part of the reasoning might have been that currently not all OCaml programs use signal handlers, so this would expose more programs than before to the fact that arbitrary OCaml code is susceptible to run inside `caml_enter_blocking_section` (in the middle of C code, when the FFI had just been changed to make sure this does not happen for allocations done from C). (At the time my own solution was to propose a `caml_enter_blocking_section_exn` variant that let the programmers clean-up resources should an exception arises.) My preference is to avoid `caml_process_pending_actions` here by conservatism but I would like to propose it for 5.1 if you think this is harmless (@kayceesrk?).

Also, this patch makes it obvious that there is a race whereby thread 2 on the same domain can reset young limit for thread 1 and make thread 1 miss a signal before it enters a blocking system call. This race was already there in OCaml 4 with `signals_are_pending` (which could be reset despite some signals still pending, otherwise OCaml 4 would have suffered from something similar to #12253). A solution would instead be to keep checking `caml_check_pending_signals()` but by comparing it with the current sigmask. But I worry that caching locally the sigmask for efficient comparison might open another can of worms, and this will remain intrinsically racy anyway.

The last solution is to proceed directly with the present PR to fix #12253.

kayceesrk commented:
Hi @gadmm, wondered if you had an answer to this question: https://github.com/ocaml/ocaml/pull/11307#pullrequestreview-1482688834. I'm keen to get this PR completed and across the line.

gadmm commented:
We are on the same page about getting this PR completed soon. This might have to wait a bit that I am back from holidays to address all the new comments.

gadmm commented:
> Some notes for self: Instead of interrupting every domain, can we not interrupt just the initial domain? We know that the initial domain will always be around. If the initial domain terminates, the program terminates. The advantage is that every domain is not penalised for the arrival of a signal. The downside is that the thread that services the initial domain may be blocked in the OS. It would make ctrl-C to be delayed arbitrarily delayed. This isn't acceptable.

IMO there are other good reasons for interrupting other domains than the initial domains, e.g. when the goal is to interrupt other domains than the initial domain. This includes terminating other domains when the initial domain is stuck inside `Domain.join` before terminating, to ensure clean program termination.

gadmm commented:
> Why shouldn't `caml_thread_yield` raise an exception? Can you clarify?

It can be called from the runtime in places where it is not safe to raise an exception or run OCaml code, after the tick thread sets `requested_external_interrupt`.

We have the choice here to make `requested_external_interrupt` a delayed interrupt able to run OCaml code like OCaml signal handlers (this was the case in OCaml 4 when implemented with signals). In this case we move the handling of `requested_external_interrupt` elsewhere and we keep `process_pending_actions`. Let me know what you think.

gadmm commented:
> If RET_FROM_C_CALL is going to be retained, it needs to be implemented for RISC-V, s390x and the WIP Power backend.

The original plan was to follow this PR with a testcase that would fail if `RET_FROM_C_CALL` is not implemented, and add instructions for backend implementers to implement `RET_FROM_C_CALL` (meant to be easy for any non-expert in possession of a godbolt).

Naturally, new backends were revived since, so they must be implemented. I still intend to write this testcase, and I can look into implementing the back-ends at the same time, but since this is a task that is easily decoupled from the current PR and that could in principle be done by anyone, I still prefer to leave it to a subsequent PR.

kayceesrk commented:
> The original plan was to follow this PR with a testcase that would fail if RET_FROM_C_CALL is not implemented, and add instructions for backend implementers to implement RET_FROM_C_CALL (meant to be easy for any non-expert in possession of a godbolt).

It would be useful to just have the testcase first and then continue the discussion from there. For this PR, the only blocking item is the introduction of `RET_FROM_C_CALL`. `RET_FROM_C_CALL` was also @xavierleroy's concern wrt performance. In order to move this PR forward, this test case will help. Without the testcase, I can't see why we should merge this PR with `RET_FROM_C_CALL`.  

> It can be called from the runtime in places where it is not safe to raise an exception or run OCaml code, after the tick thread sets requested_external_interrupt.

This makes sense to me.

> We have the choice here to make requested_external_interrupt a delayed interrupt able to run OCaml code like OCaml signal handlers (this was the case in OCaml 4 when implemented with signals). In this case we move the handling of requested_external_interrupt elsewhere and we keep process_pending_actions. Let me know what you think.

Are there downsides to doing this?

gadmm committed:
Reintroduce testcase and make it test the optimal polling

gadmm commented:
- #11596 is in fact such a testcase, with the improvement I proposed there.

- I don't see downsides for executing `requested_external_interrupt` only at places where async callbacks are executed. An upside of it is that it becomes possible to reason, inside C code, about locations where OCaml code can run concurrently on the current domain, in the same way as one can reason about where exceptions may arise. This can be useful for memory safety. So I am in favour of making the change. However, this should be considered a multicore regression better fixed separately from this PR, IMO. The question for this PR is whether we remove `caml_process_pending_actions` to make it safe for now, or we leave the original code in place in the hope that the behaviour is fixed soon enough.

gadmm committed:
[minor] Additional comments

gadmm committed:
Simplify caml_reset_young_limit

Omit a redundant call to caml_set_action_pending

gadmm committed:
Reimplement polling for signals only, in caml_enter_blocking_section

This depends on the "Optimal polling" commit

gadmm commented:
The following improvements to the "Optimal polling" commit have been made:
- Updated and clarified commit log, clarified documentation comment. Comments that were unrelated to this commit have been moved to the right commits (probably a rebase mistake).
- Implemented the Power, RISC-V and s390x back-ends. I came up with a C source that lets me copy-paste the output of godbolt almost as-is, which simplifies both writing and reviewing the RET_FROM_C_CALL macro. More details inside the commit. (Precheck will tell how it goes, I only have access to amd64 currently.)
- The testcase from #11596 has been added temporarily, for CI purposes.
- A change that was delicate to convince oneself of has been moved to a separate (and optional) commit ("Simplify caml_reset_young_limit"). I came up with an intermediate step that should be easier to understand in the context of the "Optimal polling" commit. (But I can squash it back if you were already convinced by the change in this commit.)
- This PR, and the runtime implementation in general until #12342, were careful to only run signal handlers inside `caml_enter_blocking_section`. This change still depends on the "Optimal polling" commit, but it can now be made into a separate commit ("Reimplement polling for signals only..."). The main concern of this change is to preserve the hypothesis that `caml_enter_blocking_section` never raises exceptions in programs that do not use signals to interrupt computation (which proved to be a sensitive point in the past).

There is little new code (except the power, RISC-V and s390x back-ends), but the "Optimal polling" commit should be easier to review.

gadmm commented:
In the meanwhile I was looking for more portable ways to update `young_limit` when going back to OCaml. I wondered if this was not possible with the slight cost of a function call. I ended up unconvinced, because one still has to do some dance with registers as follows:
```c
FUNCTION(G(caml_c_call))
        […]
    /* Call the function (address in %rax) */
        C_call  (*%rax)
>       movq    %rax, %r15 /* save return value */
>       movq    %r14, %rdi /* pass domain state as argument */
>       C_call  (ret_from_c_call)
>       movq    %r15, %rax /* restore return value */
    /* Prepare for return to OCaml */
        movq    Caml_state(young_ptr), %r15
    /* Load ocaml stack and restore global variables */
        SWITCH_C_TO_OCAML
        ret
```
which is barely simpler than the RET_FROM_C_CALL macro, for a likely measurable performance cost. So RET_FROM_C_CALL remains the best solution.

gadmm commented:
Precheck is green (https://ci.inria.fr/ocaml/job/precheck/864/). Only RISC-V remains untested.
The CI failure is due to the flaky test (finaliser_handover).

kayceesrk commented on Changes:
```diff
@@ -105,6 +105,17 @@ Working version
 
+- #11307: Finish implementing quality treatment for asynchronous
+  actions in multicore: soundness, liveness, and performance issues.
+  Do not wait that other threads have processed signals not for us in
```
Can you rephrase this sentence? 

> Do not wait that other threads have processed signals not for us in [caml_enter_blocking_section].

I don't understand this.

gadmm replied:
I will actually remove the line. It alludes to the fix for inifinte loop at #12342 which is already in 5.1.

kayceesrk commented on testsuite/tests/c-api/alloc_async.ml:
```diff
@@ -1,19 +1,26 @@
+   allocations from C, but runs at the first polling location in OCaml
+   code after that.
+
+   See in particular RET_FROM_C_CALL from runtime/amd64.S.
```
Is `RET_FROM_CALL` _necessary_ for this test to pass? Can you please clarify?

gadmm replied:
Yes, the test now checks that callbacks run at the first polling location in OCaml code, and the test fails by disabling RET_FROM_C_CALL. I will see if I can clarify.

Note that the intention was to include this as part of #11596, but it is possible to merge it with the present PR and close #11596.

kayceesrk replied:
I wouldn’t mind the text being part of this PR.

gadmm replied:
Ok, it is simpler this way

kayceesrk commented:
I tested the PR on RISC-V qemu docker container and all the tests pass.

kayceesrk commented:
Please can you fix the conflicts in Changes?

gadmm committed:
Reviewer feedback: improve comment

gadmm commented:
Rebased and cleaned-up history.

xavierleroy commented on runtime/power.S:
```diff
@@ -288,6 +288,15 @@ ENDFUNCTION caml_call_gc
+.macro RET_FROM_C_CALL
+        lbz     TMP, Caml_state(action_pending)
+        cmplwi  TMP, 0
+        beqlr   0
```
Bonus point for using `beqlr` (conditional branch to the link register), but we can mark this branch as (very) likely taken, it might help some POWER processors with the branch prediction.
```suggestion
        beqlr+  0
```

gadmm replied:
I thank the clang programmers for this. Neither clang nor gcc choose `beqlr+` despite the `__builtin_expect`, do you know why they make this choice?

gadmm committed:
Use beqlr+ on Power

kayceesrk commented:
Thanks @gadmm for persisting with the PR for more than a year!

gadmm commented:
Thanks!
